baybe/recommenders/base.py

```diff
@@ -50,60 +44,178 @@ class Recommender(ABC, RecommenderProtocol):
+        continuous search spaces, this method is used as a fallback for those spaces
+        as well.
+
+        Args:
+            searchspace: The hybrid search space from which to generate
+                recommendations.
+            candidates_comp: The computational representation of all discrete candidate
+                points to be considered.
+            batch_size: The size of the recommendation batch.
+
+        Raises:
+            NotImplementedError: If the function is not implemented by the child class.
+
+        Returns:
+            A dataframe containing the recommendations as individual rows.
+        """
+        raise NotImplementedError("Hybrid recommendation is not implemented.")
+
+    def _select_candidates_and_recommend(
+        self,
+        subspace_discrete: SubspaceDiscrete,
+        batch_size: int,
+    ) -> pd.DataFrame:
+        """Get candidates in a discrete search space and generate recommendations.
 
         Args:
-            searchspace: The search space.
-            recommend_discrete: The Callable representing the discrete recommendation
-                function.
-            batch_size: The chosen batch size.
-            allow_repeated_recommendations: Allow to make recommendations that were
-                already recommended earlier.
-            allow_recommending_already_measured: Allow to output recommendations that
-                were measured previously.
+            subspace_discrete: The discrete subspace from which to generate
+                recommendations.
+            batch_size: The size of the recommendation batch.
 
         Returns:
-            The recommendation in experimental representation.
+            A dataframe containing the recommendations as individual rows.
 
         Raises:
-            NotEnoughPointsLeftError: If there are fewer than ``batch_size`` points
-                left for potential recommendation.
+            NotEnoughPointsLeftError: If there are fewer points left for potential
+                recommendation than requested.
         """
-        # IMPROVE: See if the there is a more elegant way to share this functionality
-        #   among all purely discrete recommenders (without introducing complicates
-        #   class hierarchies).
-
-        # Get discrete candidates. The metadata flags are ignored if the search space
-        # has a continuous component.
-        _, candidates_comp = searchspace.discrete.get_candidates(
-            allow_repeated_recommendations=allow_repeated_recommendations
-            or not searchspace.continuous.is_empty,
-            allow_recommending_already_measured=allow_recommending_already_measured
-            or not searchspace.continuous.is_empty,
+        # Get discrete candidates
+        _, candidates_comp = subspace_discrete.get_candidates(
+            allow_repeated_recommendations=self.allow_repeated_recommendations,
```

Comment:
Can you elaborate? Not sure if I understand what exactly we are talking about

Reply:
Oh, you mean you need searchspace in order to determine whether there exists a continuous component?

Reply:
yes

Reply:
Ok, tricky one. I'm not yet sure if the current solution is ✨**ideal**✨, but I think it improves upon the previous one in several regards that I can explain. But please sit down, think about it critically, and let me know.

Here my rationale:
* What I found suboptimal about the previous one is that exactly what you described: the "local" methods (i.e. that optimize there respective search space type) still needed to know a bigger context and hence where couple to searchspace. For example: the task of `_recommend_discrete` should really be to recommend discrete spaces only – that's why we introduced the methods in the first place – but it still needed to bother about the continuous part, and vice versa.
* By not having the candidate pre-selection in the base class, it would mean that EVERY discrete recommender we would ever implement would need to conduct the same basic step first, basically by copying that snippet everywhere.
* Because the selection happened inside the methods, they had to **locally** override the flags, which was quite confusing for the coder, and in fact it took me quite a while to wrap my head around how this stuff can be even disentangled correctly. The new layout is crisp clear to the reader: the base class does the separation / candidate selection and passes them to the subroutines, i.e. there is only one place where you need to bother.
* Therefore, no more complicated passing around and overriding of flags.
* Finally, in terms of signatures, it simply makes sense to me that `_recommend_hybrid` must expect at least the union of the arguments of both the other two methods (or even a superset), because you should be able to implement the two local parts also in the hybrid function if you wanted to, at least theoretically. Thus, if we pass `candidates_comp` to the discrete one, hybrid should get it as well.
* I didn't like that `_select_candidates_and_recommend` was dangling around as a separate method and had a complicate interface where you need to pass a callable etc. Now, it is really bound to the discrete part only and can cleanly read what it needs from directly from the attributes.

Overall, the logic is quite consistent to me. However, I'm not saying that one could not change the interfaces any further. Perhaps even more simplification is possible.

What do you think?

Reply:
ok I wasnt questioning the enriety of the PR - it mostly makes sense to me. Im not sure why you gave such a all-encompassing answer summarizing all of the PR - it makes me think there might be more to my question than anticipated, but perhaps not

- is it right that `_select_candidates_and_recommend` is only ever used for purely discrete spaces? This answers my question in that a distinction in the flags is obsolete

But this has now opened the following confusion:
- this question arose because previously the docstring of `_select_candidates_and_recommend` says it will also be used for hybrid spaces, hence the old distinction in the flags
- however, having looked at it it seems that was not really the case, at least not everywhere that function was used and sometimes hybrid recommenders go directly via `discrete.get_candidates`
- Potential problem: this would mean that metadata for the discrete part in hybrid spaces is not marked. If true that was a problem even before this PR but the problem has been transferred now to a central location

Reply:
To your first question: yes, that is the idea.

For the rest: also yes. Where which part was used and via what interface was really not standardized in the old version. And correct: the discrete part is not marked in hybrid spaces, and it wasn't before either. This is partly due to the fact that we haven't really ever talked about in detail what a desired/optimal behavior would be in hybrid spaces. But also I tend to think it really has no purpose in discrete spaces. However, this is something we can (and should!) definitely discuss – how about on Tuesday? Could you put in on our discussion list, so that we don't forget? But because it wasn't handled cleanly before this PR, we can probably also neglect it here and bother in a later PR?

Reply:
pretty sure this was discussed before, in fact I think that marking was one of the main points of the old `_select_candidates_and_recommend` and addition to just calling `.get_candidates` - it was not correctly implemented though, probably forgotten

the real question is whether the marking is needed for hybrid spaces. I don't think it hurts to have the info of which subspace parts have been visited - especially if the logic to mark is identical as in purely discrete case

since the issue was there before im fine with neglecting in this PR, my original point was also answered